service/matching/taskListManager.go

```diff
@@ -42,20 +42,95 @@ const (
 	done time.Duration = -1
 )
 
+// NOTE: Is this good enough for stress tests?
+const (
+	_maxDispatchDefault = 100000.0
+	_dispatchLimitTTL   = time.Second
+)
+
 type taskListManager interface {
 	Start() error
 	Stop()
 	AddTask(execution *s.WorkflowExecution, taskInfo *persistence.TaskInfo) error
 	GetTaskContext(ctx context.Context) (*taskContext, error)
 	SyncMatchQueryTask(ctx context.Context, queryTask *queryTaskInfo) error
 	CancelPoller(pollerID string)
+	UpdateMaxDispatch(maxDispatchPerSecond *float64)
 	String() string
 }
 
-func newTaskListManager(e *matchingEngineImpl, taskList *taskListID, config *Config) taskListManager {
+type rateLimiter struct {
+	maxDispatchPerSecond *float64
+	tokenBucket          common.TokenBucket
+	ttl                  *time.Timer
+	sync.RWMutex
+	ttlRaw time.Duration
+}
+
+func newRateLimiter(maxDispatchPerSecond *float64, ttl time.Duration) rateLimiter {
+	return rateLimiter{
+		maxDispatchPerSecond: maxDispatchPerSecond,
+		tokenBucket: common.NewTokenBucket(
+			int(*maxDispatchPerSecond), common.NewRealTimeSource(),
+		),
+		ttlRaw: ttl,
+		ttl:    time.NewTimer(ttl),
+	}
+}
+
+func (rl *rateLimiter) UpdateMaxDispatch(maxDispatchPerSecond *float64) {
+	if rl.shouldUpdate(maxDispatchPerSecond) {
+		rl.Lock()
+		defer rl.Unlock()
+		rl.maxDispatchPerSecond = maxDispatchPerSecond
+		rl.tokenBucket = common.NewTokenBucket(int(*maxDispatchPerSecond), common.NewRealTimeSource())
+	}
+}
+
+func (rl *rateLimiter) shouldUpdate(maxDispatchPerSecond *float64) bool {
+	rl.RLock()
+	defer rl.RUnlock()
+	if maxDispatchPerSecond == nil {
```

Comment:
No need to put this check under the lock.

Reply:
Makes sense